testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,907 @@
+    stream.start(mockClientStreamListener);
+    client.shutdown();
+    ClientTransport.PingCallback mockPingCallback = mock(ClientTransport.PingCallback.class);
+    try {
+      client.ping(mockPingCallback, MoreExecutors.directExecutor());
+    } catch (UnsupportedOperationException ex) {
+      assumeTrue(false);
+    }
+    verify(mockPingCallback, timeout(1000)).onSuccess(Matchers.anyInt());
+    stream.cancel(Status.CANCELLED);
+  }
+
+  @Test
+  public void ping_afterTermination() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    client.shutdown();
+    ClientTransport.PingCallback mockPingCallback = mock(ClientTransport.PingCallback.class);
+    try {
+      client.ping(mockPingCallback, MoreExecutors.directExecutor());
+    } catch (UnsupportedOperationException ex) {
+      assumeTrue(false);
+    }
+    verify(mockPingCallback, timeout(1000)).onFailure(throwableCaptor.capture());
+    Status status = Status.fromThrowable(throwableCaptor.getValue());
+    assertCodeEquals(Status.UNAVAILABLE, status);
+  }
+
+  @Test
+  public void newStream_duringShutdown() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    // Stream prevents termination
+    ClientStream stream = client.newStream(methodDescriptor, new Metadata());
+    stream.start(mockClientStreamListener);
+    client.shutdown();
+    ClientStream stream2 = client.newStream(methodDescriptor, new Metadata());
+    ClientStreamListener mockClientStreamListener2 = mock(ClientStreamListener.class);
+    stream2.start(mockClientStreamListener2);
+    verify(mockClientStreamListener2, timeout(1000))
+        .closed(statusCaptor.capture(), any(Metadata.class));
+    assertCodeEquals(Status.UNAVAILABLE, statusCaptor.getValue());
+
+    // Make sure earlier stream works.
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(1000, TimeUnit.MILLISECONDS);
+    serverStreamCreation.stream.close(Status.OK, new Metadata());
+    verify(mockClientStreamListener, timeout(1000))
+        .closed(statusCaptor.capture(), any(Metadata.class));
+    assertCodeEquals(Status.OK, statusCaptor.getValue());
+  }
+
+  @Test
+  public void basicStream() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    Metadata clientHeaders = new Metadata();
+    clientHeaders.put(asciiKey, "client");
+    clientHeaders.put(asciiKey, "dupvalue");
+    clientHeaders.put(asciiKey, "dupvalue");
+    clientHeaders.put(binaryKey, "äbinaryclient");
```

Comment:
:smile: Über is my typical test string, but it didn't align with how my values were descriptive.